internal/server/singleprocess/state/track_task.go

```diff
@@ -0,0 +1,317 @@
+package state
+
+import (
+	"strings"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/hashicorp/go-memdb"
+	bolt "go.etcd.io/bbolt"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	pb "github.com/hashicorp/waypoint/pkg/server/gen"
+)
+
+var tracktaskBucket = []byte("tracktask")
+
+func init() {
+	dbBuckets = append(dbBuckets, tracktaskBucket)
+	dbIndexers = append(dbIndexers, (*State).tracktaskIndexInit)
+	schemas = append(schemas, tracktaskIndexSchema)
+}
+
+// TrackTaskPut creates or updates the given TrackTask.
+func (s *State) TrackTaskPut(t *pb.TrackTask) error {
+	memTxn := s.inmem.Txn(true)
+	defer memTxn.Abort()
+
+	err := s.db.Update(func(dbTxn *bolt.Tx) error {
+		if t.TaskJob == nil {
```

Comment:
Actually thinking maybe this doesn't need to be required, depending on at what point we have the server upsert a TrackTask. I _think_ the first place will be where we wrap a requested job with Start/Stop tasks and I think by that point we might know the job ids for everything - So maybe this is ok. The downside if there's other task data to update later (say in the runner) the requester will need to make sure to include the original TaskJob Id along with the original TrackTask Id.

Maybe something to leave for now - and fix once the server actually uses this state?

Reply:
I'd keep it as an assertion. I think you're right: we'll always have the job.